### ISSUE TITLE
more detailed instructions on usage of response data and purpose of generated classes

### ISSUE THREAD

**Issue by xpavp03** [phpfui/ConstantContact#23]
Hi Bruce,
thank you for this library. Following the documentation I managed to query Constant Contact for a list of campaigns but I'm struggling to understand how I'm supposed to work with the response.

The output from \PHPFUI\ConstantContact\V3\Emails::get() is a nested array of arrays. But I see a ton of classes generated from Swagger in PHPFUI\ConstantContact\Definition. Why is the response not mapped to these objects?

In my case, when I fetch a list of e-mail campains why don't I get an instance of \PHPFUI\ConstantContact\Definition\PagedEmailCampaignResponse?

By trial and error I found that I can easily hydrate it myself but I sense that there must be a reason for why it doesn't happen automatically:
```php
    $arrayResponse = $campaignsClient->get();
    $response = new PagedEmailCampaignResponse($arrayResponse);
```

I found in other tickets that you suggest looking at your Bicycle Club website. You use the response as an array. Why don't you use \PHPFUI\ConstantContact\Definition\Contacts when fetching contacts?
https://github.com/phpfui/BicycleClubWebsite2023/blob/4a4efd4848f44c3b811776e1ed72de01bad520d8/App/Cron/Job/ConstantContactSync.php#L52

I'm puzzled but still appreciative of the time and efforts you've put into this library. Thank you!

**Comment by phpfui** [comment]
This is a very good question. Probably because I did not see that the responses were typed when I first did this, as there was a lot to do just to get something working.

But typed responses are a good idea. I think I can add get methods to return the defined schemas. Will probably look like getReturnSchema, postReturnSchema so it does not conflict with the normal array return of the base method.

Will look into what is involved with that.  Thanks for the suggestion.

**Comment by phpfui** [comment]
OK, I added ReturnSchema methods for all operations, ie. getReturnSchema() will return a hydrated object, get() will continue to return a raw array.

Change your version to "dev-main" in composer.json and them "composer update"

Let me know if you find any issues.  Once you test it, I will publish a new release.

**Comment by xpavp03** [comment]
I also spent some time thinking about the naming convention.
`getReturnSchema()` to me sounds like it will return meta-data about the schema of the return value.

Alternative names could be `getTyped()` or `getObject()`.
Also, perhaps including `getArray()` too, with `get()` calling `getArray()` for backward compatibility. Possibly with `/** @deprecated */` should you ever decide to remove `get()` to force the user to be explicit.

Or to have only one method `get()` with an additional parameter `$returnType` that would accept a constant `ReturnType::Array` or `ReturnType::Object`. Kind of like Doctrine `getResult($hydrationMode = self::HYDRATE_OBJECT)`, or Nette `Container::getValues()`.

Default for `$returnType` would be null and if not specified, `$client->defaultReturnType` would be used. Or `ReturnType::Array` if no `$client->defaultReturnType` was specified. That way `get()` would continue to return arrays unless told otherwise.

**Comment by xpavp03** [comment]
My first test was with expired token which brought up the question of what should `getReturnSchema()` return in case of an error.

`Client::get()` returns an empty array which may make it difficult to distinguish success, from failure. I don't know if CC ever returns an empty response.

I see two possible alternative paths:

1. both `Client::get()` and `getReturnSchema()` return null on errors

2. throw a custom exception
containing the Guzzle exception message and response status code, replacing the `$client->lastError` and `$client->statusCode`
`Client::get()` continues to always return array (now happening only on success) and `getReturnSchema()` continues to return an object

(2) seems cleaner to me at the cost of a BC break
(1) is a BC break only if the user didn't check for errors using `$client->lastError` or `$client->statusCode`, or using `empty($list)`.

**Comment by phpfui** [comment]
I'll try to address everything in one comment.

* Namespaces. I probably should have put this in my PHPFUI namespace, I just did not think of that when I developed it, as I don't really see this as my library (but it is), but just a wrapper around Constant Contact.  To change this, I would have to abandon this version and create a new version with a different versioning schema to allow for bigger changes.  Right now the versioning tracks the Constant Contact API revs.  It would need to track my revs as well, so 1.24.10.1 for example.
* I can't really change the namespaces for the definitions because the CC API does not have those namespaces. All this code is autogenerated from the YAML file and it would be hard to parse that in any consistent pattern.  I will look into a Response namespace, as I think I can break that out. Will check.
* I can look into generators. I have not used them much since 5.5 as I don't think they add much to code readability.
* If I go with a Response namespace, then I think getResponse(), postResponse(), etc. would be the way to go.  Actually you don't really need an array return type at all as the Definition base class as a getData() method.  Should probably be called toArray() though (this would match my ORM library and seems to be PHP native). Easy to add.
* I think I should return null on error. Good idea. Not a fan of exceptions for possible expected behavior.

Should have something by end of my day or next.

Thanks for testing and the input!

**Comment by phpfui** [comment]
I took a look at the uniqueness of Response types, and they are not unique in all cases, so I left the Response types mixed in with the standard definitions.

I changed the method suffix to Typed() when they return an object.

I now return null on error. Not sure how much of a breaking change this is. Will see if it is problem.

**Comment by phpfui** [comment]
Yes, I wanted to put responses in a separate namespace to show they were really different, but the schema has some responses that are also used as input, so I decided to leave things alone, since the schema does not separate them.

As for exceptions vs return values, some modern languages like GO do not use exceptions. Some shops restrict their use as well. In my years of software development, I have often found they are used around every method call, and therefore don't really simplify the code. I tend to want to keep processing things, to get as much done as possible. Not processing an entire batch of orders if one is bad is not what most people want or expect. You handle the issue, and continue to process.

But I do understand the use of exceptions, and if you want to abandon your entire process, exceptions are for sure easier to deal with.

In your case, I would implement it this way:
```php
    $campaignsResponse = $campaignsClient->getTyped();

    if ($this->testForError('message', $campaignsClient)) {
      return [];
    }

    do {
      foreach ($campaignsResponse->campaigns ?? [] as $campaignStub) {
...
        $campaignDetailResponse = $campaignClient->getTyped($campaignStub->campaign_id);
        
        foreach ($campaignDetailResponse->campaign_activities ?? [] as $activityStub) {
          $activityResponse = $activitiesClient->getTyped($activityStub->campaign_activity_id, 'permalink_url');
          if ($this->testForError('message', $activitiesClient)) continue;
         }
         $this->testForError('message', $campaignsClient);
       }
       $this->testForError('message', $campaignsClient);

...
```
I do throw exceptions for type errors, just like a compiler would. I have just seen too many examples of exception abuse, that I don't make them standard operating procedures.

And thanks for letting me know about typed returns. I missed that when I first implemented things.

**Comment by xpavp03** [comment]
Thank you for the explanation, I understand.

Please let me know if I can help with anything here, or we can close this ticket.
